### ISSUE TITLE
filter-api socketpair leak?

### ISSUE THREAD

**Issue by bdijkstra82** [OpenSMTPD/OpenSMTPD#674]
After some hours or days after starting smtpd I get:
 
Apr 13 21:39:58 belisama spf[25177]: warn: filter-api:spf socketpair: Too many open files
Apr 13 21:39:58 belisama delay[14001]: warn: session e6640a1435e43218: write out of sequence
Apr 13 21:39:58 belisama last message repeated 95 times
 
After a first such warning, the same happens on every incoming message. The filter process then has several hundred sockets to its name.
 
The same has also occurred with a filter chain of only filter-pause but this didn't cause any
real problems. However, with my filter-spf in the chain this results in messages being accepted with an empty data portion, which is highly undesirable.
 
~~On an error, socketpair() sets errno, so as a work-around I think I should reject with 451 if errno is found to be nonzero inside a filter callback. Should this work?~~ Probably not.

Using `fstat -u _smtpd|wc -l` I can monitor the amount of open files. With just one filter:
```
filter delay pause
listen on egress filter delay 
```
the amount of open files (at apparent "idle" state) seems to grow gradually.

Apr 15 10:17:09 belisama smtpd[375]: info: OpenSMTPD 5.9.1 starting
output: 67
Apr 18 10:11:34
output: 145
About 500 counts of `smtp-in: New session` happened during that time.

**Comment by bdijkstra82** [comment]
After a few hours of running with `-dv -Tfilters -Tsmtp` it seems that some sessions never finish the EOM sequence. Final IO_DATAIN still happens (presumably containing QUIT command), but IO_DISCONNECT doesn't occur, last trace of such a session looks like:
`filter: filter_tx_io: datain (765) for req a31f7330e940add3`
after which a 'healthy' session would continue with:
`filter: imsg IMSG_FILTER_RESPONSE from procfilter delay[hooks=0xffffffff,flags=0x0000]`
etc.

The corresponding TCP sockets are kept in CLOSE_WAIT forever. 

I'm not 100% sure, but it seems that these 'hanging' sessions contain messages that have been accepted but will never be delivered (because stopping the server deletes the ramqueue). That would be Really Bad(TM) so for now I'm not going to run any filters on my production server.

**Comment by poolpOrg** [comment]
We're going to investigate and fix this shortly.

Please do keep in mind that filters are enabled to provide users with a way to test them and report failures like you did, but they are considered experimental as stated in the release note from OpenBSD 5.7.1:

    Experimental: 
        * filters plumbing is now always enabled
        * the filter API is now available, though considered experimental

Please don't use them in production :-)

**Comment by bdijkstra82** [comment]
Yeah I guess I was a little too eager to test my code in the wild. :-)

Looking forward to the fix.

**Comment by ericfaurot** [comment]
The filter-pause is very unfair in that it sleep()s on any new connection, suspending all event processing. This certainly triggers some weird behaviors in the chain. Investigating.

**Comment by poolpOrg** [comment]
On Wed, Apr 27, 2016 at 08:35:24AM -0700, Eric Faurot wrote:
> The filter-pause is very unfair in that it sleep()s on any new connection, suspending all event processing. This certainly triggers some weird behaviors in the chain. Investigating.
> 

I discussed this with Joerg Jung a few days ago and told him the same,
he has sent us a diff to introduce an async timer to sleep the session
and not the filter, it's pending review.

However, we should still investigate the "weird behaviors" so this
ticket should remain open disregarding the filter-pause design issue

-- 
Gilles Chehade

https://www.poolp.org                                          @poolpOrg

**Comment by bdijkstra82** [comment]
One can expect filters to take time in processing, wait for I/O, etc. Whether that takes 5ns or 5s, that shouldn't cause incorrect behaviour. So yes, please investigate.

**Comment by poolpOrg** [comment]
The leak issue is being investigated by @ericfaurot, however:

> This certainly triggers some weird behaviors in the chain.

That comment will remain valid, we have an async non-blocking design and filters have to be written with this in mind, so if a filter has a design issue causing it to block, the entire filter chain is blocked and smtpd will not work around that.

When the leak issue is fixed, filter-pause will still need to be fixed as it is completely broken as is. Joerg has sent a diff pending review for that matter.

**Comment by ericfaurot** [comment]
So I found an issue in the code that explains "out of sequence" lines. The attached diff fixes this, but I couldn't reproduce the leak. Can you run with that diff and let us know how it goes?
[fdout.diff.txt](https://github.com/OpenSMTPD/OpenSMTPD/files/240199/fdout.diff.txt)

**Comment by poolpOrg** [comment]
@bdijkstra82 care to give us a feedback ? :-)

**Comment by bdijkstra82** [comment]
I'm currently pointing blacklisted IPs to a chrooted smtpd which runs a filter that pauses for 1 second on every event and then rejects the message at EOM. Running for two hours now, haven't seen a socket leak yet.

**Comment by bdijkstra82** [comment]
@ericfaurot "couldn't reproduce" - did you run smtpd with filter-pause for many hours and did you feed it many (overlapping) conversations with varying client behaviour?

After my previous message, I ran that setup for a day or so **without** the patch, to make sure that socket leaks happen and continue to happen. They do. **With** the patch, socket leaks also happen, but I haven't hit the open file limit yet.

Also, both with and without the patch, I noticed that the session hangs after my filter gives a 451 rejection. The connection stays open, but no response comes after a command (e.g. QUIT). The trace just says `<<< [MSG] QUIT`. After this, some clients hang up, but many stay until the 5-minute time-out. I'll let smtpd.conf reject the messages and test again to see if this is a separate issue.
